### ISSUE TITLE
Switch off plots

### ISSUE THREAD

**Issue by JanaKoehler1985** [mzechmeister/viper#31]
Maybe we should enter an option to switch off all the plots generated during the modelling process. Even they are useful for the developer and some interested users to check if everything is going right, they are also very time consuming. And if you already know everything is fine and you just one to run viper again changing just one parameter (like kapsig ), it is not necessary to watch all the plots again.

I'm just thinking about how to do it best. In the moment the plots are always generated when kwargs are given in S.mod.fit (like dx, rel_fac,...):

https://github.com/mzechmeister/viper/blob/f5424531dbf13c0d9502c00c47752ae8474d98a1/model.py#L141-L142

So whether we set all the related parameters to false or would it be better to add a parameter for the plotting options instead of using kwargs?

**Comment by mzechmeister** [comment]
Yeah, I agree.
As a command line argument it could be a new or overloaded `look` option.
Probably best is to set a new default for `look`.

I think now, having `show` inside `.fit` is a bit unexpected, unusual and difficult to understand, though convenient.
Not sure if this might be a case for a decorator or defined as a wrapper in `viper.py`.

I also often uncomment this:
https://github.com/mzechmeister/viper/blob/f5424531dbf13c0d9502c00c47752ae8474d98a1/viper.py#L559

**Comment by JanaKoehler1985** [comment]
Oh right, we have a  `look` option. But in the moment this is just used to pause the plots but not to switch it off. It also could be an idea to chance `look` -> `showplots` (or something similar) and than have the options: `off` (no plots), `look` (run through plots) and `pause` (show plots and pause).

> I think now, having show inside .fit is a bit unexpected, unusual and difficult to understand, though convenient.

Good point. Even so it doesn't disturb me that much. The comment you posted of cause also have to be considered when switching of the plots.

**Comment by JanaKoehler1985** [comment]
While talking about the look option, I am not sure I like the way we call it by now:
https://github.com/mzechmeister/viper/blob/95b553ef0c9e188f6b70b0c9bcda3e8b28b5457f/viper.py#L422
https://github.com/mzechmeister/viper/blob/95b553ef0c9e188f6b70b0c9bcda3e8b28b5457f/viper.py#L367

What is the meaning of `o` here? Havn't changed it during the renaming of the variables to `order` as I was not sure about it.

**Comment by mzechmeister** [comment]
Yes, `o` is just `order`.

**Comment by mzechmeister** [comment]
I have no strong opinion yet and I'm hesitating.
I'm quite used to the `look` naming, albeit, I can understand if the naming sounds strange to other.
`showplots` is a bit too long/verbose to me. Just `plot` might be an alternative.

`-look none` might used to switch off plot, thus overloading `look`.

**Comment by JanaKoehler1985** [comment]
> Yes, o is just order.

Ok, so it really is order and still confusing. Why not just `if look` or `if lookguess`? Why is it related to the order?

**Comment by JanaKoehler1985** [comment]
Right, `look none` and also `look pause` sounds good and I am fine with keeping look. I just had no idea for a name when showing the plots without pause. Like, I guess, we use it most time.

**Comment by mzechmeister** [comment]
`look` can be e.g. `13,17` or `2:5`, so only specific orders can be shown/paused.

(`look` is currently a global variable; actually the boolean should be passed as an argument in `fitchunk(..., makeplot=order in look)`, but it introduce mores variables.)

**Comment by JanaKoehler1985** [comment]
Ok, good to know. This was not clear to me and conflicts with my ideas above :-/

**Comment by JanaKoehler1985** [comment]
Or we do it similar to our possible new `kapsig`  option, like `look pause 3:7`. And by using just `look pause` all orders are used. For `none` no further input is needed anyhow. 
Still would have no idea for just running through the plots without pause.... maybe `look show`? (not really nice either)

**Comment by mzechmeister** [comment]
I had the same in mind. But I think, `pause` in `-look` is probably too much. It mixes slice and boolean variables.
And it complicates argument parsing currently `type=arg2range`.

Also it needs to be decided, if `pause` or `nopause` should be default.
serval basically works like `pause=True`; `pause=false` cannot be set.

**Comment by JanaKoehler1985** [comment]
I see. So how about `-look` (plot all) , `-look 3:5` (plot selected orders), `-look none` (switch off plotting plots; does it work with slice type?). And than we take `-pause` as extra parameter to pause and set default to none?  

(or `-look 0` to switch off)

**Comment by mzechmeister** [comment]
> `-look 0`

Unfortunately, Python indexing is zero-based.

Another suggestion:
`-look` as now
`-show` same as `look`, but without pause

As default nothing should be shown.

`-lookguess`, `-lookpar`, `-lookres` will have no `show` counterpart.

So we end up with an additional option :/ + variable (as required).

Coding ideas:
short and lazy
```
if look: show = look

if o in show:
    gplot ...
    if look: pause()    
```

"cleaner" interface and with more variables (whose names might be improved...)
```
if look: show = look

fitchunk(..., make_plot=o in show, plot_pause=bool(look))
if make_plot: 
    gplot ...
    if plot_pause: pause()
```

**Comment by JanaKoehler1985** [comment]
Ok... but if we are thinking about adding an additional option anyhow: Wouldn't it be easier to leave everything as it is and add an option `-show` (or other name), which we set to False as default. And if it is selected as true we use it in model.py by replacing kwargs :
```
 if kwargs: 
     self.show(params, pixel, spec_obs, par_rv=par_rv, **kwargs) 
```

and for the one plotting line in viper:

`gplot+(pixel[flag_obs != 0], wave_obs[flag_obs != 0], spec_obs[flag_obs != 0], 1*(flag_obs[flag_obs != 0] == flag.clip), 'us (lam?$2:$1):3:(int($4)?5:9) w p pt 6 ps 0.5 lc var t "flagged and clipped"') `

**Comment by mzechmeister** [comment]
On command line I would strongly prefer to type just `-look 3,4,6` instead of `-look 3,4,6 -show`.

Internally, the variables might transform, as we like. We could also make `look` the master (as now):
```
if show: look = show

if look:
    ...
    if not show: pause()
```
But to me the suggestion above causes less head aches.

**Comment by JanaKoehler1985** [comment]
Just to make sure, I got it right. So it will be whether the `-look` or `-show` option which is chosen by the user. If none of both is chosen, no plotting at all. Well, indeed would be an option.

But wouldn't it then be better to use `-look` for showing the plots and something like `-lookstop` (sounds stupid too) for showing and pause? I don't know, look and show just sound so similar. You know what I mean?

**Comment by mzechmeister** [comment]
> Just to make sure, I got it right. So it will be whether the `-look` or `-show` option which is chosen by the user. If none of both is chosen, no plotting at all. Well, indeed would be an option.

Yes.

> But wouldn't it then be better to use `-look` for showing the plots and something like `-lookstop` (sounds stupid too) for showing and pause?

If you find a better name... (or more active for show: looknonstop, movie, film, watch)

> I don't know, look and show just sound so similar. You know what I mean?

I know, but it was acceptable for me.

**Comment by JanaKoehler1985** [comment]
Finding names was never my strength either. Nevertheless, I see it like that: In the moment the most important thing is to have an option to run viper without plotting all outputs. Therefore I am fine with using `look` and `show` (or looknonstop? lookfast? lookrun?)
(And if we ever find a better name, we still can change)

Do you prefer to implement these changes yourself (whenever you have time) to make sure everything is how you prefer? Plus, I guess have you have a better understanding of the current plotting options ;-)

**Comment by JanaKoehler1985** [comment]
Ok, I was thinking about it again and may found a solution. (Or a new starting point for discussion).

If I see it right, we want 3 plotting options:

1. Fast plotting       - via `lookfast` (or other name)
2. Plot with pause - via `look` 
3. No plotting     - Default;` look = None`, `lookfast = 0` 

Based on that we can do
```
if order in look:
    show = 1 
else:
    show = lookfast

params, e_params = S_mod.fit(pixel_ok, spec_obs_ok, par_rv, par_norm, 
par_wave_guess, par_ip, par_atm, par_bkg=par_bkg, parfix_bkg=parfix_bkg, 
dx=0.1*show, sig=sig[i_ok], res=(not createtpl)*show, rel_fac=createtpl*show)

if show:
        # overplot flagged and clipped data
        gplot+(pixel[flag_obs != 0], wave_obs[flag_obs != 0], spec_obs[flag_obs != 0], 
1*(flag_obs[flag_obs != 0] == flag.clip), 'us (lam?$2:$1):3:(int($4)?5:9) w p pt 6 ps 0.5 lc 
var t "flagged and clipped"')
```

And in `model.py` it would be:

 ```
 if res or rel_fac or dx:        
        gplot(*args)
```

**Comment by mzechmeister** [comment]
Implemented in ab44796cbcb979ec6b56ae72c7e9c76da011542e.
